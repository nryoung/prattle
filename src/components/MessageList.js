import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import withChatkit from '../containers/WithChatkit';
import media from '../styles/media';
import Message from './Message';

const StyledMessageList = styled.div`
  display: flex;
  flex-wrap: wrap;
  overflow-y: scroll;
  padding: 72px 16px 72px;

  ${media.screenMediumUp`
    width: 600px;
    margin: 0 auto;
  `}

  // Add this to hide scrollbars on chrome
  // when using a non standard mouse on mac
  &::-webkit-scrollbar {
    width: 0;
  }
`;

class MessageList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentUserId: null,
      messages: [],
    };
    this.lastMessage = React.createRef();
  }

  scrollToBottom = () => {
    window.scrollTo(0, this.lastMessage.current.offsetTop);
  };

  componentDidMount() {
    const { chatkit } = this.props;
    chatkit.connect().then((currentUser) => {
      currentUser.subscribeToRoom({
        roomId: process.env.REACT_APP_ROOM,
        hooks: {
          onMessage: (message) => {
            this.setState(
              {
                currentUserId: currentUser.id,
                messages: [...this.state.messages, message],
              },
              () => this.scrollToBottom()
            );
          },
        },
      });
    });
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  render() {
    const { messages, currentUserId } = this.state;
    return (
      <StyledMessageList>
        {messages.map((message) => (
          <Message
            key={message.id}
            message={message}
            isOwnMessage={currentUserId === message.sender.id}
          />
        ))}
        <div ref={this.lastMessage} />
      </StyledMessageList>
    );
  }
}

MessageList.propTypes = {
  chatkit: PropTypes.object.isRequired,
};

export { MessageList };

export default withChatkit(MessageList);
