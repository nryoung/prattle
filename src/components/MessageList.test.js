import React from 'react';
import renderer from 'react-test-renderer';
import MessageList from './MessageList';

const props = {
  messages: [
    {
      id: ':messageId',
      userId: ':userId',
      text: ':text',
    },
  ],
};

it('it renders', () => {
  const component = renderer.create(<MessageList {...props} />);
  expect(component.toJSON()).toMatchSnapshot();
});
