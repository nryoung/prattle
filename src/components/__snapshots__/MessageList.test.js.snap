// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div
  className="sc-ifAKCX eDrjbq"
>
  <div
    className="sc-bdVaJa BWkgA"
  >
    <div
      className="sc-bwzfXH haoZKa"
    >
      :userId
      :
    </div>
    <div
      className="sc-htpNat LIIZh"
    >
      <span
        className="sc-bxivhb Igthg"
      >
        :text
      </span>
    </div>
  </div>
</div>
`;
